Ubernetes Lite support for GCE

justinsb opened the pull request:
This adapts the GCE cloudprovider to work across multiple zones.

Posting a WIP so we have something concrete to discuss.

We will need to resolve how scoping works when names are ambiguous.  My understanding is that it is currently permitted to have two volumes both named "myvolume" in two different zones; we will need some policy for how we resolve a volume reference named "myvolume" in this case.  I believe something similar is planned to instance name scoping also.

justinsb committed:
GCE: Allow for reuse of master

This is for internal use at the moment, for testing Ubernetes Lite, but
arguably makes the code a little cleaner.

Also rename KUBE_SHARE_MASTER -> KUBE_USE_EXISTING_MASTER

justinsb committed:
Ubernetes Lite GCE: Support multiple zones in GCE cloud provider

We adapt the existing code to work across all zones in a region.

We require a feature-flag to enable Ubernetes-Lite

Reasons:

* There are some behavioural changes if users create volumes with
the same name in two zones.
* We don't want to make one API call per zone if we're not running
Ubernetes-Lite.
* Ubernetes-Lite is still experimental.

There isn't a parallel flag implemented for AWS, because at the moment
there would be no behaviour changes from this.

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -65,7 +65,9 @@ type GCECloud struct {
-	zone              string
+	region            string
+	instanceZone      string
+	managedZones      []string
```
The meanings of these two zone constructs are not clear.  Please comment to clarify.

justinsb replied:
I believe you're taking over this PR, do you want me to add these comments?

In case it isn't clear...
instanceZone = the zone which this instance is in
managedZones = all zones which kubernetes is managing (i.e. all zones in the region for Ubernetes-Lite, just the current zone if we're not running Ubernetes-Lite)

quinton-hoole replied:
Now that you've taken the PR back again, yes please :-)

roberthbailey replied:
How do you dynamically add zones to this list after the cluster is running? The way you explained creating an ubernetes lite cluster at the SIG meeting was that you'd set up a cluster normally and then re-run the kube-up script adding additional zones. That doesn't seem to jive with pre-propulating a fixed list of zones that are being "managed". 

@brendandburns

roberthbailey replied:
Can the "managed zones" be dynamically inferred from the location of the nodes that have registered with the cluster?

a-robinson replied:
That would be ideal, so that switching a cluster between single zone and multi-zone (or vice versa) wouldn't require restarting the controller manager.

justinsb commented:
I added a feature-flag for GCE, because of the behavioral change around volumes when they have the same name, and because we don't want to be scanning multiple zones when we're not running Ubernetes-Lite.

Example of how to try this out:
```
MULTIZONE=1 KUBE_GCE_ZONE=us-central1-a cluster/kube-up.sh
KUBE_SHARE_MASTER=true MULTIZONE=1 KUBE_GCE_ZONE=us-central1-b cluster/kube-up.sh
KUBE_SHARE_MASTER=true MULTIZONE=1 KUBE_GCE_ZONE=us-central1-f cluster/kube-up.sh

kubectl get nodes

cd examples/guestbook-go; find *.json | xargs -I {} kubectl create -f {}

kubectl scale rc guestbook --replicas=12

sleep 20 # Wait for scaling and for the load-balancer to be created

kubectl describe service guestbook

# Get the load-balancer IP

(for i in `seq 20`; do curl -s http://<ip>:3000/env | grep HOSTNAME; done)  | sort | uniq | wc -l


# When you're finished
KUBE_SHARE_MASTER=true KUBE_GCE_ZONE=us-central1-b cluster/kube-down.sh
KUBE_SHARE_MASTER=true KUBE_GCE_ZONE=us-central1-f cluster/kube-down.sh
KUBE_GCE_ZONE=us-central1-a cluster/kube-down.sh
```

I still need to tag GCE volumes; that is next up.

justinsb commented:
Removing WIP from title :-)

quinton-hoole commented:
Flaky unit test tracked in #19268 . Try again.

@k8s-bot unit test this

quinton-hoole commented on cluster/gce/util.sh:
```diff
@@ -589,6 +589,22 @@ function kube-up {
   find-release-tars
   upload-server-tars
 
+  if [[ ${KUBE_SHARE_MASTER:-} == "true" ]]; then
```
bikeshedding type nit: KUBE_USE_EXISTING_MASTER rather?

roberthbailey replied:
Your refactoring and @fgrzadkowski's refactoring in #19218 look to have some overlap and some conflicts. We need to decide which one to merge first.

justinsb replied:
I have no objection to #19218 merging first (happy to rebase); we both see the need to modularize this code.  I made the minimally invasive change, but #19218 is much better.

But if #19218 isn't going to merge soon...

roberthbailey replied:
I think that @fgrzadkowski may be OOO for a while so let's not hold this up on his PR.

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -145,6 +148,23 @@ func getNetworkNameViaAPICall(svc *compute.Service, projectID string) (string, e
+		return nil, fmt.Errorf("unexpected response listing zones: %v", err)
+	}
+	zones := []string{}
+	for _, zone := range res.Items {
```
nit: It seems that the API can do this filtering for you:

https://github.com/google/google-api-go-client/blob/master/compute/v1/compute-api.json#L13925

Less code is better :-)

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -76,6 +78,7 @@ type Config struct {
 		TokenBody   string `gcfg:"token-body"`
 		ProjectID   string `gcfg:"project-id"`
 		NetworkName string `gcfg:"network-name"`
+		Multizone   string `gcfg:"multizone"`
```
This needs to be a boolean, not a string?

justinsb replied:
My intention here was to allow additional behaviours in future.  e.g. directly specifying the managed zones, rather than managing all the zones in a region.  That said, if we can auto-configure based on node registration, this requirement goes away.  And maybe it should just be a bool anyway for now, and we can add a list of zones if/when we need it.

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -178,15 +206,19 @@ func newGCECloud(config io.Reader) (*GCECloud, error) {
 		if cfg.Global.TokenURL != "" {
 			tokenSource = newAltTokenSource(cfg.Global.TokenURL, cfg.Global.TokenBody)
 		}
+		if cfg.Global.Multizone != "" {
```
Related to above comment, this config param needs to be a boolean. As it stands, unless I'm mistaken, setting MULTIZONE=false or MULTIZONE=0 will be equivalent to setting MULTIZONE=true, which is downright confusing.

justinsb replied:
OK, I'll make it a bool :-)

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -220,11 +252,23 @@ func CreateGCECloud(projectID, zone, networkURL string, tokenSource oauth2.Token
 		networkURL = gceNetworkURL(projectID, networkName)
 	}
 
+	if len(managedZones) == 0 {
```
You actually set it to nil above.

justinsb replied:
Only in the case when multizone is set though, right?  This looks correct to me.

The weird structuring comes because when we are parsing the configuration, we don't yet have a GCE client, so we can't yet query the list of zones in the region.  So we pass nil to mean "all zones in region".

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -377,11 +420,16 @@ func isHTTPErrorCode(err error, code int) bool {
-func (gce *GCECloud) EnsureTCPLoadBalancer(name, region string, requestedIP net.IP, ports []*api.ServicePort, hosts []string, affinityType api.ServiceAffinity) (*api.LoadBalancerStatus, error) {
-	if len(hosts) == 0 {
+func (gce *GCECloud) EnsureTCPLoadBalancer(name, region string, requestedIP net.IP, ports []*api.ServicePort, hostNames []string, affinityType api.ServiceAffinity) (*api.LoadBalancerStatus, error) {
+	if len(hostNames) == 0 {
```
I don't think that this or the next diff belong in this PR?

quinton-hoole replied:
Oh, never mind, I see why lower down.

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -725,32 +773,41 @@ func (gce *GCECloud) firewallObject(name, region, ipAddress string, ports []*api
-	// Add the fields we want
-	listCall = listCall.Fields("items(name,tags)")
+	for zone, hostNames := range hostNamesByZone {
+		listCall := gce.service.Instances.List(gce.projectID, zone)
```
indentation?

justinsb replied:
I don't understand.  I think this is gofmt-ed...

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1648,6 +1740,84 @@ func (gce *GCECloud) ListClusters() ([]string, error) {
+	Kind string
+}
+
+func (gce *GCECloud) resolveHosts(hostNames []string) ([]*gceInstance, error) {
```
The use of "resolve" and "hostName" makes this sound like it's doing DNS resolution, when  it's not.  
Prefer to stick with the original naming scheme (  getInstanceByName(name) ) ?

justinsb replied:
Good call.  I went with getInstanceByName and getInstance*s*ByName*s* (note the pluralizations in the multi-fetch method)

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1648,6 +1740,84 @@ func (gce *GCECloud) ListClusters() ([]string, error) {
+	return hosts, nil
+}
+
+func (gce *GCECloud) resolveHost(name string) (*gceInstance, error) {
```
Same comment as above.

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1427,31 +1475,38 @@ func (gce *GCECloud) ExternalID(instance string) (string, error) {
 // InstanceID returns the cloud provider ID of the specified instance.
 func (gce *GCECloud) InstanceID(instance string) (string, error) {
-	return gce.projectID + "/" + gce.zone + "/" + canonicalizeInstanceName(instance), nil
+	host, err := gce.resolveHost(instance)
```
Please be consistent with the use of 'host' and 'instance' here. I prefer instance, although I understand the potential name clash problem that perhaps you were trying to solve.  How about:

```
instanceObj, err := gce.getInstanceByName(instance)
```
?

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1427,31 +1475,38 @@ func (gce *GCECloud) ExternalID(instance string) (string, error) {
 	var instances []string
-	for _, instance := range res.Items {
-		instances = append(instances, instance.Name)
+	// TODO: Parallel?  (But no generics seems to make this harder)
```
Agreed that it could be improved, although it's not clear how generics would help you here?  Given that the loop is O(zones) I'm OK with leaving parallelization as a todo.

justinsb replied:
Removed snarky note about generics :-)

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1533,40 +1591,49 @@ func (gce *GCECloud) DeleteRoute(clusterName string, route *cloudprovider.Route)
+// Create a new Persistent Disk, with the specified name & size.
+// If zone is specified (!= ""), the disk will be created in that zone.
+// If the zone is not specified (""), the disk will be created in the zone in which we are currently running.
+// Currently, this will be the zone where the kube-controller-manager is running.
```
Actually, more technically correct, where this code is currently running, I think.  In theory, it could be running anywhere (e.g. if called by kubelet).

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1533,40 +1591,49 @@ func (gce *GCECloud) DeleteRoute(clusterName string, route *cloudprovider.Route)
 		SizeGb: sizeGb,
 	}
-	createOp, err := gce.service.Disks.Insert(gce.projectID, gce.zone, diskToCreate).Do()
+	if zone == "" {
```
I really don't like an empty string here.  Ideally get the caller to pass in the sone name, or perhaps nil, if that's not feasible.

justinsb replied:
Moved it to the caller; also helps with the previous comment

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1609,15 +1681,35 @@ func (gce *GCECloud) DiskIsAttached(diskName, instanceID string) (bool, error) {
 
-func (gce *GCECloud) getDisk(diskName string) (*compute.Disk, error) {
-	return gce.service.Disks.Get(gce.projectID, gce.zone, diskName).Do()
+func (gce *GCECloud) resolveDisk(diskName string) (*gceDisk, error) {
```
Prefer getDiskByName()

quinton-hoole replied:
And as per previous comments, I think that we can always easily pass in a zone here too, and get around the problem that you mention.

justinsb replied:
Agree with getDiskByName naming.

The instance suggestion is clever - I'll do it :-)

justinsb replied:
DeleteDisk is problematic, unfortunately, so I don't think we can entirely eliminate the zone scanning.

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1533,40 +1591,49 @@ func (gce *GCECloud) DeleteRoute(clusterName string, route *cloudprovider.Route)
-func (gce *GCECloud) CreateDisk(name string, sizeGb int64) error {
+// Create a new Persistent Disk, with the specified name & size.
+// If zone is specified (!= ""), the disk will be created in that zone.
+// If the zone is not specified (""), the disk will be created in the zone in which we are currently running.
```
Perhaps this defaulting could be applied throughout to resolve the name clash problem below?

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1533,40 +1591,49 @@ func (gce *GCECloud) DeleteRoute(clusterName string, route *cloudprovider.Route)
 
 func (gce *GCECloud) DeleteDisk(diskToDelete string) error {
-	deleteOp, err := gce.service.Disks.Delete(gce.projectID, gce.zone, diskToDelete).Do()
+	disk, err := gce.resolveDisk(diskToDelete)
```
I would suggest making all of the Disk() methods consistent, and take a zone parameter.

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1533,40 +1591,49 @@ func (gce *GCECloud) DeleteRoute(clusterName string, route *cloudprovider.Route)
 
 func (gce *GCECloud) AttachDisk(diskName, instanceID string, readOnly bool) error {
-	disk, err := gce.getDisk(diskName)
+	disk, err := gce.resolveDisk(diskName)
```
Here you can get the zone from the instance, as it only makes sense to attach them if they're in the same zone.

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1626,18 +1718,18 @@ func getGceRegion(zone string) (string, error) {
-func (gce *GCECloud) ListClusters() ([]string, error) {
-	list, err := gce.containerService.Projects.Zones.Clusters.List(gce.projectID, gce.zone).Do()
+func (gce *GCECloud) listClustersInZone(zone string) ([]string, error) {
+	list, err := gce.containerService.Projects.Zones.Clusters.List(gce.projectID, zone).Do()
```
Whooah!  This actually seems to call into GKE to list clusters?  Seems that it does not belong in the GCE interface here?  No your fault, at all, but surprising, nonetheless.

quinton-hoole replied:
@roberthbailey @a-robinson any idea what's going on here?

roberthbailey replied:
@brendandburns should have a good answer since he added it in https://github.com/kubernetes/kubernetes/pull/2355.

justinsb replied:
I kinda assumed this was for internal use by the GKE code?  It isn't used anywhere that *I* can see...

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1648,6 +1740,84 @@ func (gce *GCECloud) ListClusters() ([]string, error) {
 	return result, nil
 }
 
+func (gce *GCECloud) ListClusters() ([]string, error) {
```
See above.  It looks like we might be able to delete this entire thing, or just return nil (like AWS and all other cloud providers seem to do).

quinton-hoole commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1648,6 +1740,84 @@ func (gce *GCECloud) ListClusters() ([]string, error) {
+	//  * check each zone in turn (in parallel?)
+	//  * try local zone first (so we don't pay the tax unless we have to)
+	//  * remove instances from the list once we've found them
+	//  * we could cache 'last known zone' by name, but I don't know if this will be needed
```
Yes, I think we need to improve this, as right now it's O(instances x zones) in the number or remote calls it makes (we could have thousands of instances, so that's real bad).

A pretty simple improvement to reduce it to O(zones) would be to switch around resolveHost() to call resolveHosts() rather than the other way around.  The heavy work would get done by resolveHosts (which would look very much like the current resolveHost(), iterating over all zones, getting all matching instances from each of them (with a server-side filter).

quinton-hoole replied:
We can defer parallelizing, removing instances once they're found, caches etc until later, as I don't think they'll really buy us that much anyway.

quinton-hoole commented on pkg/volume/gce_pd/gce_util.go:
```diff
@@ -137,7 +137,7 @@ func (gceutil *GCEDiskUtil) CreateVolume(c *gcePersistentDiskProvisioner) (volum
 	// GCE works with gigabytes, convert to GiB with rounding up
 	requestGB := volume.RoundUpSize(requestBytes, 1024*1024*1024)
-	err = cloud.CreateDisk(name, int64(requestGB))
+	err = cloud.CreateDisk(name, "", int64(requestGB))
```
As per previous comment, I think you can pass in cloud.GetZone() instead of the ugly empty string here.

quinton-hoole commented:
@justinsb LGTM barring comments. Let me know what you think, or once you've addressed them.

Q

quinton-hoole commented:
@justinsb Also needs a rebase now.

roberthbailey commented:
@quinton-hoole - I'm concerned about the way that this PR seems to require that you know apriori which zones are going to be managed by the cluster. /cc @a-robinson

quinton-hoole commented:
@roberthbailey 
```
// If managedZones is nil / empty all zones in the region will be managed.
```
managedZones is just a way to know where to look for zonal resources (e.g. instances, volumes) if they're not found in the current zone).

Make sense?

@justinsb this still needs a rebase.

roberthbailey commented:
Does this mean it isn't possible to manage zones from other regions? If I specify an explicit list of zones that are in a different region from the master will they be managed?

quinton-hoole commented:
@roberthbailey There is no way to pass an explicit list of zones into this script, as far as I can see.  managedZones is a local variable, initialized to all of the zones in the same region as the current zone being built.  The intention is that all managed zones are in the same region.  That's fundamental to Ubernetes Lite.  If they're not in the same region, then various things will not work as intended (including load balancing).   Make sense?

justinsb commented:
Working through the changes here, but the big one seems to be how we feature-flag it.

I think it needs a feature flag, because otherwise the behaviour of Kubernetes when there are two volumes with the same name in different zones changes, when Ubernetes-Lite is not being used.

Right now, you have to set the feature flag when you are first creating the master (i.e. the first kube-up).  Then the only available behaviour is that *all* zones in the region are managed.  Then you add nodes sharing the same master, in different zones (in the same region).

Given we have to feature flag it anyway, I think this is probably acceptable.  I think it would be nice to skip zones in which we know we have no instances, but I'm thinking we can get more gains by optimizing getInstancesByNames.

I like the suggestion of watching nodes when they are registered, and then automatically enabling Ubernetes-Lite when you register nodes in a second zone.  I guess we could do this using an AdmissionController.  I'm just not sure how much it gains us.  I don't think we can do it in the cloud controller, because it relies on Node objects (which we don't see).

Of course, if we can figure out dynamic cluster reconfiguration in general, I would be very happy!

justinsb commented:
Rebased & pushed code addressing code review.  I rewrote getInstancesByNames so it will be O(zones) instead of O(zones * instances)

Along with the feature-flag issue (as described in my previous comment), there's a slight change in that some functions now check that the instance exists and make an extra API call to do so.  I could tweak it so that we don't do that in the single zone case, but looking at the uses it doesn't seem worthwhile.

I've kept the commits separate for easier reviews; I'll have to squash before the (eventual) merge.

justinsb commented:
@k8s-bot unit test this

quinton-hoole commented:
@justinsb Awesome!  Thanks for addressing all the comments.  LGTM!
Let me know as soon as you've squashed the changes, and I will add the LGTM label to have merge-bot merge it.

The unit/integration test failure is a bit weird.  The log only says the following:

FAIL	k8s.io/kubernetes/pkg/cloudprovider/providers/gce [build failed]

But if that was the case, how on earth did the e2e tests pass?  Assuming some sort of strange flake, I'll try again.  But suffice to say, that will need to pass before we can merge.

@k8s-bot unit test this

justinsb commented:
I think/hope the build fail was because I hadn't updated a field name in the tests (localZone).  Squashed to two sensible commits, and will hopefully pass tests shortly.

quinton-hoole commented:
\o/  LGTM!  Label added.

quinton-hoole commented:
@justinsb The e2e test failure looks legit.  Minor script error by the looks of things.  Probably best to give it a quick test-run before re-submitting.  Let me know when you do so that I can re-add the LGTM label.

```
13:05:10 /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce@5/kubernetes/hack/e2e-internal/../../cluster/../cluster/../cluster/gce/util.sh: line 699: fail: unbound variable
13:05:10 2016/01/20 21:05:10 e2e.go:309: Error running up: exit status 1
13:05:10 2016/01/20 21:05:10 e2e.go:305: Step 'up' finished in 43.851535285s
13:05:10 2016/01/20 21:05:10 e2e.go:141: Error starting e2e cluster. Aborting.
```

justinsb commented:
Oops sorry.  I will re-test.

justinsb commented:
The root cause of the e2e test failure was this I believe:
```
ERROR: (gcloud.compute.instances.create) Some requests did not succeed:
 - Quota 'IN_USE_ADDRESSES' exceeded.  Limit: 65.0
```

This was reported as `...cluster/gce/util.sh: line 699: fail: unbound variable` because of an error in wait-for-jobs: fail is not exported or set when a job fails.  Opened #19908

However, this made it clear to me that in fact I hadn't correctly threaded through the MULTIZONE flag, and also that zone selection wasn't working.  I changed the log level so multi-zone configuration is always reported i.e. V(2) -> Info; I changed the zone discovery so it goes back to the dumb way and doesn't try to use a filter (sorry), and I made sure MULTIZONE was threaded through correctly.

I hope we can get those e2e test running against this soon ... they definitely beat any testing it is practical for me to do.  And I'll get to work on tagging the volumes so we can have volume zone-awareness working with GCE too (should be relatively simple).

alex-mohr commented:
@k8s-bot test this

quinton-hoole commented:
@justinsb Thanks!

>  I changed the zone discovery so it goes back to the dumb way and doesn't try to use a filter (sorry)

Why?  Just curious.

justinsb commented:
@quinton-hoole I couldn't get a filter on region to work; it wasn't returning any zones.  I also tried it in the console https://cloud.google.com/compute/docs/reference/latest/zones/list and couldn't get it to return values with a filter.  I tried both with short-form region (us-central-1) and long form (https://....)  If you can get it to work, I'd be very happy to put it back, but you'll have to tell me the secret!

a-robinson commented:
Not that it's important, but when you were testing it, did you use us-central-1 or us-central1?

Using the filter `name eq us-central1-.` works for me, e.g. `GET https://www.googleapis.com/compute/v1/projects/arob-k8s/zones?filter=name+eq+us-central1-.`

quinton-hoole commented:
@justinsb ^^
Sounds fine for now.  We can re-add the filter in a subsequent PR if we solve the quandry.

justinsb commented:
Ah!  I was setting the filter to `region eq us-central1`  or `region eq http://....`  I didn't get an error - I just got an empty list.  The name thing is a good idea though it does embed an assumption about the name structure of regions.

And yeah, I always put the hypen in the wrong place the first time, but I think I tried both ways when testing :-)

yujuhong commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1663,6 +1783,129 @@ func (gce *GCECloud) ListClusters() ([]string, error) {
+		zone := gce.managedZones[0]
+		res, err := gce.service.Instances.Get(gce.projectID, zone, name).Do()
+		if err != nil {
+			if !isHTTPErrorCode(err, http.StatusNotFound) {
```
If `isHTTPErrorCode(err, http.StatusNotFound)` is true, line 1886 below will try to access `res.Zone`, causing nil-pointer dereference.

yujuhong commented:
kube-controller-manager.log from kubernetes-e2e-gce-slow/1534:
```
I0121 21:08:28.895874       6 nodecontroller.go:745] node e2e-slow-minion-guc6 hasn't been updated for 45.256594055s. Last out of disk condition is: &{Type:OutOfDisk Status:Unknown LastHeartbeatTime:2016-01-21 21:07:42 +0000 UTC LastTransitionTime:2016-01-21 21:08:23 +0000 UTC Reason:NodeStatusUnknown Message:Kubelet stopped posting node status.}
E0121 21:08:28.988419       6 util.go:76] Recovered from panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/util.go:70
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/util.go:58
/usr/src/go/src/runtime/asm_amd64.s:402
/usr/src/go/src/runtime/panic.go:387
/usr/src/go/src/runtime/panic.go:42
/usr/src/go/src/runtime/sigpanic_unix.go:26
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/gce/gce.go:1886
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/gce/gce.go:1497
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/node/nodecontroller.go:508
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/node/nodecontroller.go:232
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/util.go:120
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/util.go:121
/usr/src/go/src/runtime/asm_amd64.s:2232
```

justinsb commented:
Ouch ... will fix right away

gmarek commented:
This was a cause of #20788
